pennylane_lightning/_serialize.py

```diff
@@ -121,13 +132,50 @@ def _serialize_hamiltonian(ob, wires_map: dict, use_csingle: bool):
     return hamiltonian_obs(coeffs, terms)
 
 
+def _serialize_pauli_word(ob, wires_map: dict, use_csingle: bool):
+    """Serialize a PauliWord into a Named or Tensor."""
+    if use_csingle:
+        named_obs = NamedObsC64
+        tensor_obs = TensorProdObsC64
+    else:
+        named_obs = NamedObsC128
+        tensor_obs = TensorProdObsC128
+
+    if len(ob) == 1:
+        wire, pauli = list(ob.items())[0]
+        return named_obs(pauli_name_map[pauli], [wires_map[wire]])
+
+    return tensor_obs(
+        [named_obs(pauli_name_map[pauli], [wires_map[wire]]) for wire, pauli in ob.items()]
+    )
+
+
+def _serialize_pauli_sentence(ob, wires_map: dict, use_csingle: bool):
+    """Serialize a PauliSentence into a Hamiltonian."""
+    if use_csingle:
+        rtype = np.float32
+        hamiltonian_obs = HamiltonianC64
+    else:
+        rtype = np.float64
+        hamiltonian_obs = HamiltonianC128
+
+    pwords, coeffs = zip(*ob.items())
+    terms = [_serialize_pauli_word(pw, wires_map, use_csingle) for pw in pwords]
+    coeffs = np.array(coeffs).astype(rtype)
+    return hamiltonian_obs(coeffs, terms)
+
+
 def _serialize_ob(ob, wires_map, use_csingle):
     if isinstance(ob, Tensor):
         return _serialize_tensor_ob(ob, wires_map, use_csingle)
     elif ob.name == "Hamiltonian":
         return _serialize_hamiltonian(ob, wires_map, use_csingle)
+    elif _obs_has_kernel(ob):
+        return _serialize_named_obs(ob, wires_map, use_csingle)
+    elif ob._pauli_rep is not None:
+        return _serialize_pauli_sentence(ob._pauli_rep, wires_map, use_csingle)
```

Comment:
I think we need to switch the order of these two.

Reply:
Hi @timmysilv, I transferred the logic you are introducing here to [my PR](https://github.com/PennyLaneAI/pennylane-lightning/pull/425).
What I observed, and discussed with @albi3ro, was that the operator S_Prod was reaching the method `_serialize_named_obs`, changing the order really helped with that.
It is possible that I made a mistake in the process, so let me know your thoughts on that.

Reply:
I switched the order locally to demonstrate what I'm trying to avoid:
```pycon
>>> import pennylane as qml
>>> from pennylane_lightning._serialize import _serialize_ob
>>> _serialize_ob(qml.PauliZ(0), wires_map={0: 0}, use_csingle=True)
Hamiltonian: { 'coeffs' : [1], 'observables' : [PauliZ[0]]}
```

I am surprised that an SProd returned true for `obs_has_kernel`, but it is kinda the thing I was scared of. I checked locally, and it returns `True` when you import `is_pauli_word` from `pennylane.pauli`, but lightning is currently importing the deprecated one from `pennylane.grouping`. Anyway, this clearly needs some fixing because both approaches cause problems. I'll separate things out and make it better

Reply:
I came here to say what you said in your edit :sweat_smile:.
But yes, Lightning is still importing `is_pauli_word` from `pennylane.grouping`, but in my PR I already update it to the `pennylane.pauli`. 
Thank you for your help with that. 
As soon as you have a more definite solution I will update it in my PR.

Reply:
Is this resolved?

Reply:
I believe so. I've summarized the relevant changes [in this comment](https://github.com/PennyLaneAI/pennylane-lightning/pull/424#discussion_r1171848775), and added test cases to ensure things work as expected